java-frontend/src/test/java/org/sonar/java/ast/JavaAstScannerTest.java

```diff
@@ -481,54 +486,53 @@ public void scanFile(JavaFileScannerContext context) {
 
   @Test
   void test_should_fail_on_stackoverflow_by_default() {
-    InputFile trivialCompilationUnit = TestUtils.inputFile("src/test/resources/AstScannerNoParseError.txt");
-    var files = List.of(trivialCompilationUnit);
-    var problematicVisitor = new JavaFileScanner(){
-      @Override
-      public void scanFile(JavaFileScannerContext context) {
-        throw new StackOverflowError();
-      }
-    };
-    List<JavaFileScanner> visitors = List.of(problematicVisitor);
+    var files = List.of(TRIVIAL_COMPILATION_UNIT);
+    List<JavaFileScanner> visitors = List.of(PROBLEMATIC_VISITOR);
 
     // Assert that by default, StackOverflowError is propagated
     var emptySettings = new MapSettings();
-    assertThrows(StackOverflowError.class, () ->
-      scanFilesWithVisitorsAndContext(files, visitors, emptySettings, JavaVersionImpl.MAX_SUPPORTED));
+    assertThrows(StackOverflowError.class, () -> scanFilesWithVisitorsAndContext(files, visitors, emptySettings));
   }
 
   @Test
   void test_should_not_fail_on_stackoverflow_when_set() {
-    InputFile trivialCompilationUnit = TestUtils.inputFile("src/test/resources/AstScannerNoParseError.txt");
-    var files = List.of(trivialCompilationUnit);
-    var problematicVisitor = new JavaFileScanner(){
-      @Override
-      public void scanFile(JavaFileScannerContext context) {
-        throw new StackOverflowError();
-      }
-    };
-    List<JavaFileScanner> visitors = List.of(problematicVisitor);
-
     // Assert that when configured to not fail on exception, StackOverflowError is swallowed, but logged
     MapSettings settings = new MapSettings().setProperty(SonarComponents.SONAR_FAIL_ON_STACKOVERFLOW, false);
-    assertDoesNotThrow(() -> scanFilesWithVisitorsAndContext(files, visitors, settings, JavaVersionImpl.MAX_SUPPORTED));
+    assertDoesNotThrow(() -> scanFilesWithVisitorsAndContext(List.of(TRIVIAL_COMPILATION_UNIT), List.of(PROBLEMATIC_VISITOR), settings));
     assertThat(logTester.logs(Level.ERROR))
       .containsExactly("A stack overflow error occurred while analyzing file: 'src/test/resources/AstScannerNoParseError.txt'");
   }
 
+  @ParameterizedTest
+  @MethodSource("failFastAndFailOnSOCombinations")
+  void assertFailFastWithFailOnStackOverflowInteraction(boolean failFast, boolean failOnSO, boolean shouldFail) {
+    MapSettings settings = new MapSettings()
+      .setProperty(SonarComponents.FAIL_ON_EXCEPTION_KEY, failFast)
+      .setProperty(SonarComponents.SONAR_FAIL_ON_STACKOVERFLOW, failOnSO);
+    var files = List.of(TRIVIAL_COMPILATION_UNIT);
+    List<JavaFileScanner> visitors = List.of(PROBLEMATIC_VISITOR);
+
+    if (shouldFail) {
+      assertThrows(StackOverflowError.class, () -> scanFilesWithVisitorsAndContext(files, visitors, settings));
+    } else {
+      assertDoesNotThrow(() -> scanFilesWithVisitorsAndContext(files, visitors, settings));
+    }
+  }
+
+  static List<Object[]> failFastAndFailOnSOCombinations() {
```

Comment:
No need to shorten it here
```suggestion
  static List<Object[]> failFastAndFailOnStackOverflowCombinations() {
```